[InstCombine] Handle more scalable geps in EmitGEPOffset

davemgreen opened the pull request:
Following up on #71565, this makes scalable splats in EmitGEPOffset use the ElementCount as opposed to assuming it is fixed width, and attempts to handle scalable offsets with vector geps by splatting the vscale to each vector lane. This meant performing the mul by scale separately from CreateVScale.

nikic commented on llvm/lib/Analysis/Local.cpp:
```diff
@@ -66,16 +66,26 @@ Value *llvm::emitGEPOffset(IRBuilderBase *Builder, const DataLayout &DL,
+          VScale = Builder->CreateVScale(ConstantInt::get(
+              cast<VectorType>(IntIdxTy)->getElementType(), 1));
+          VScale = Builder->CreateVectorSplat(
+              cast<VectorType>(IntIdxTy)->getElementCount(), VScale);
```
Can we instead move the above splat creation until after the multiply? I.e. calculate on scalars and then splat?

davemgreen commented:
It appears that the `& PtrSizeMask` can be removed without altering any of the lit tests or any of the programs I tried across AArch64/Arm. Let me know if you think I should re-add it.

davemgreen committed:
[InstCombine] Handle more scalable geps in EmitGEPOffset

Following up on #71565, this makes scalable splats in EmitGEPOffset use the
ElementCount as opposed to assuming it is fixed width, and attempts to handle
scalable offsets with vector geps by splatting the vscale to each vector lane.

It appears that the `& PtrSizeMask` can be removed without altering any of the
tests or any of the test I tried across AArch64/Arm.

nikic commented:
> It appears that the `& PtrSizeMask` can be removed without altering any of the lit tests or any of the programs I tried across AArch64/Arm. Let me know if you think I should re-add it.

I think this is okay. I confirmed that the APInt uint64_t ctor will implicitly truncate the value, rather than asserting. As such, an explicit mask should not be necessary.

nikic commented on llvm/lib/Analysis/Local.cpp:
```diff
@@ -36,25 +36,18 @@ Value *llvm::emitGEPOffset(IRBuilderBase *Builder, const DataLayout &DL,
       if (StructType *STy = GTI.getStructTypeOrNull()) {
         uint64_t OpValue = OpC->getUniqueInteger().getZExtValue();
-        Size = DL.getStructLayout(STy)->getElementOffset(OpValue);
+        unsigned Size = DL.getStructLayout(STy)->getElementOffset(OpValue);
```
```suggestion
        uint64_t Size = DL.getStructLayout(STy)->getElementOffset(OpValue);
```

nikic reviewed (approved):
LGTM

davemgreen committed:
Fixup - Whoops, had the wrong type.
